### DISCUSSION TITLE
Using Xpra for dual screen display on vGPU

### DISCUSSION THREAD

**Discussion by s-bonnet** [Xpra-org/xpra#3907] [Q&A]
Hello !
I'll try to explain my use case and maybe I can get an answer overthere :)

I've an HMI running on Azure on a vGPU (Tesla M60). In the same pod of this app we have : 
- a container running with Xorg
- a home made window manager

The configuration of Xorg is made so that there is 1 screen of 3840x1200, with 2 monitors of size 1920x1200. Everything is virtualized.

We then try to retrieve the display on local machines, with 2 physical screens (each with a resolution of 1920x1200). Obvisoulsy, to make is easier, I'm looking for a fullscreen display on each screen.

If I summarize, I've a 3840x1200 virtual screen on Azure, and I would like to get the display on 2 physical screens : 
- on the first one, resolution of 1920x1200 and start a (0,0) on the virtual screen
- on the second one, resolution of 1920x1200 and start at (1920,0) on the virtual screen

I'm trying to resolve my problem with Xpra, and getting the display in a firefox on my local machine.

First of all, do you think I can solve this problem with Xpra ? We want to try it as the performances seems really better than VNC.

For the while, what I did is : 
- start a server in another container (within the same pod) : xpra start-desktop --use-display=yes --bind-tcp=8080
- configure my firefox client with a width of 3840.

However, when i go full screen, I get the whole display on only 1 screen...

I thought to 2 different solutions :
- having the possibility to connect 2 clients, each with a resolution of 1920x1200 and one with an offset of (1920,0). But it seems that this is not possible with current options when connecting a client
- launching 2 xpra servers, each one mapping with a subpart of the virtualized screen (ie one starting at (0,0), and one starting with and offset, at (1920,0)). Then i'll connect 2 clients, 1 on each server.
Is it possible to configure such solutions in current versions of Xpra ? Or is it possible to make it another way ? Or is it simply not possible :) ?

Thanks a lot for reading and for the help!

**Comment by totaam** [comment]
> The configuration of Xorg is made so that there is 1 screen of 3840x1200, with 2 monitors of size 1920x1200. Everything is virtualized.

Why not virtualize the same geometry that you have on the client?
Everything would then be so much easier.

> First of all, do you think I can solve this problem with Xpra ?

Yes.
> We want to try it as the performances seems really better than VNC.

That's usually the case.

> launching 2 xpra servers, each one mapping with a subpart of the virtualized screen (ie one starting at (0,0), and one starting with and offset, at (1920,0))

That's the way I would do it.
You could start with shadow mode (slower and less efficient but easier to get going):
https://github.com/Xpra-org/xpra/issues/3384#issuecomment-1513949823
But ideally you would want to use monitor mode. It might just work.
```
xpra monitor :0 --resize-display=no
```

**Reply by s-bonnet** [reply]
Hello,

Thanks for the reply. Actually we tried the shadow mode but as we noticed drops of performances (confirmed by the documentation) we didn't continue to look on this solution.

We also tried to start the monitor mode, but it seems to launch a Xvfb. As we already have a Xorg running on the required $DISPLAY , there is a conflict and xpra does not start with this error message "Vfb failed to start, exiting" (which makes sense).

Regarding the configuration, our v-GPU has only 1 connected screen, so we couldn't configure Xorg with the configuration of 2 screens.

**Comment by totaam** [comment]
> We also tried to start the monitor mode, but it seems to launch a Xvfb.

It doesn't if you specify `--use-display=yes` or if you specify a display that is already running on the command line.

> "Vfb failed to start, exiting" (which makes sense).

My guess is that you are running an [ancient version from your distro](https://github.com/Xpra-org/xpra/wiki/Distribution-Packages) rather than a supported version.
Current versions detect that the display is running and re-use it automagically. (`use-display=auto` is now the default)

The monitor server currently detects existing Xorg displays connected and assigns a window model to each one:
https://github.com/Xpra-org/xpra/blob/ad68834183996289f93b7e4d612b9a7762e42f32/xpra/x11/desktop/monitor_server.py#L109-L113
It should be possible to:
* add a virtual display to your existing Xorg - perhaps via `evdi` (though the performance many not be as good: #3390)
* split your display into virtual ones - perhaps using [`fakexrandr`](https://github.com/phillipberndt/fakexrandr)
* teach the monitor server to parse command line arguments to split the screen

**Comment by s-bonnet** [comment]
Hello !

I'm running this version of xpra:

> xpra --version
> xpra v4.4.6-r2

You can find below some information regarding my configuration & some logs:
- Xorg is running within its own container
- Xpra is launched within another container
- Both containers run in the same pod

The xrandr configuration (launched in the Xpra container)

> root@thmi-app-nsp-monolith-f588cb6cb-k9j8q:/# xrandr --listmonitors
> Monitors: 2
>  0: D1 1920/1x1080/1+0+0  DVI-D-0
>  1: D2 1920/1x1080/1+1920+

When I start xpra in monitor mode:

> xpra start-monitor :1 --use-display=yes

The process does not start with the logs:

> 2023-07-24 08:32:12,579 no uinput module (not usually needed)
> xpra initialization error:
>  the vfb cannot virtualize monitors

If i use the desktop mode it starts, but I only get the display of the monitor :0, not the :1 

(I tried both mode, as i didn't really get the difference between monitor & desktop).

Regards,

**Reply by totaam** [reply]
> xpra v4.4.6-r2

This is not a full package name or version. See https://github.com/Xpra-org/xpra/wiki/Reporting-Bugs

> `the vfb cannot virtualize monitors`

Please try with `--debug screen,randr` to get a more detailed error message.
Perhaps we can force it to accept this type of config.

**Comment by totaam** [comment]
> `the vfb cannot virtualize monitors`

Probably a Debian based distro.
One of the key reasons why they are not fully supported is bugs such as this one: they cannot support Xdummy properly.
